contracts/exchangeIssuance/ExchangeIssuance.sol

```diff
@@ -0,0 +1,384 @@
+    address private WETH;
+
+    /* ============ Events ============ */
+
+    event ExchangeIssue(address indexed recipient, address indexed setToken, address indexed inputToken, uint256 amount);
+    event ExchangeRedeem(address indexed recipient, address indexed setToken, address indexed outputToken, uint256 amount);
+
+    /* ============ Constructor ============ */
+
+    constructor(
+        address _uniFactory,
+        IUniswapV2Router02 _uniRouter, 
+        address _sushiFactory, 
+        IUniswapV2Router02 _sushiRouter, 
+        IBasicIssuanceModule _basicIssuanceModule
+    ) 
+        public
+    {
+        uniFactory = _uniFactory;
+        uniRouter = _uniRouter;
+
+        sushiFactory = _sushiFactory;
+        sushiRouter = _sushiRouter;
+
+        WETH = uniRouter.WETH();
+        basicIssuanceModule = _basicIssuanceModule;
+        IERC20(WETH).approve(address(uniRouter), PreciseUnitMath.maxUint256());
+        IERC20(WETH).approve(address(sushiRouter), PreciseUnitMath.maxUint256());
+    }
+
+    /* ============ External Functions ============ */
+
+    receive() external payable {}
+
+    /**
+     * Runs all the necessary approval functions required before issuing
+     * or redeeming a set token. This function must only be called before the first time
+     * this smart contract is used on any particular set token, or when a new token is added
+     * to the set token.
+     *
+     * @param _setToken    Address of the set token being initialized
+     */
+    function initApprovals(ISetToken _setToken) external {
+        ISetToken.Position[] memory positions = _setToken.getPositions();
+        for (uint256 i=0; i<positions.length; i++) {
+            IERC20 token = IERC20(positions[i].component);
+            token.approve(address(uniRouter), PreciseUnitMath.maxUint256());
+            token.approve(address(sushiRouter), PreciseUnitMath.maxUint256());
+            token.approve(address(basicIssuanceModule), PreciseUnitMath.maxUint256());
+        }
+    }
+
+    /**
+     * Redeems a set token and sells the underlying tokens using Uniswap
+     * or Sushiswap.
+     *
+     * @param _setToken             Address of the set token being redeemed
+     * @param _amountSetToRedeem    The amount of the set token to redeem
+     * @param _isOutputETH          Set to true if the output token is Ether
+     * @param _outputToken          Address of output token. Ignored if _isOutputETH is true
+     * @param _minOutputReceive     Minimum amount of output token / ether to receive
+     */
+    function exchangeRedeem(ISetToken _setToken, uint256 _amountSetToRedeem, bool _isOutputETH, address _outputToken, uint256 _minOutputReceive) external nonReentrant {
+        _setToken.transferFrom(msg.sender, address(this), _amountSetToRedeem);
+        basicIssuanceModule.redeem(_setToken, _amountSetToRedeem, address(this));
+        liquidateComponents(_setToken);
+        if(_isOutputETH) {
```

Comment:
For better function encapsulation and function logic abstraction, we could put this if/else logic into an internal function called `_handleRedeemOutput`